### PR TITLE
[Chan Wai Hon Jonathan] Duke Increments

### PR DESCRIPTION
**Pull request by jonchan51**

### CODE REVIEW

File: src/main/java/Duke.java [hunk 1]
```diff
@@ -24,13 +27,18 @@ public void run() {
         sayGreeting();
```
Comment (seahlynn) [inline src/main/java/Duke.java]: Think it's great that you separate the welcome greeting and goodbye into functions, to get them 'out of the way'!

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -24,13 +27,18 @@ public void run() {
...
+            case "bye":
                 break;
+            default:
+                addTask(command);
             }
-            printFormatted(command);
-        }
+        } while (!command.equals("bye"));
         sayBye();
     }
```
Comment (seahlynn) [inline src/main/java/Duke.java]: Splitting the different lines of codes in response to different commands makes your code really readable, which I really like!! Should have done this myself :)

File: src/main/java/Duke.java [hunk 3]
```diff
@@ -24,13 +27,18 @@ public void run() {
...
+        do {
             command = sc.nextLine();
-            if (command.equals("bye")) {
+            switch (command) {
+            case "list":
+                printTasks();
+                break;
+            case "bye":
                 break;
+            default:
```
Comment (seahlynn) [inline src/main/java/Duke.java]: Would it be better to place the addTask() function under one of the cases ("add") instead of under the default clause? Because I think that following coding standards, "default:" should be used to catch any input that's "out of the norm", instead of being used to process a valid input!

**Review by seahlynn — COMMENTED** [review]
Very readable code!! Goood job!!

**Review by seahlynn — COMMENTED** [review]
I like how all your String messages are represented as static variables with very clear names!!! It unclutters your code and makes it very easy to follow!  (Sorry this is not a very constructive review but it really impressed me HAHAHA)

File: src/main/java/duke/command/FindCommand.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    public void execute(TaskList tasks, Storage storage) throws DukeException {
+        TaskList filteredTasks = new TaskList();
+        for (Task task : tasks) {
+            if (task.getDescription().contains(search)) {
+                filteredTasks.add(task);
+            }
+        }
+
+        StringBuilder lines = new StringBuilder();
```
Comment (seahlynn) [inline src/main/java/duke/command/FindCommand.java]: Maybe the naming of this variable could be clearer? Something along the lines of "toPrint" etc!

File: src/main/java/duke/command/Command.java [hunk 5]
```diff
@@ -5,8 +5,20 @@
...
+     * Executes command.
+     *
+     * @param tasks Current TaskList.
+     * @param storage Current Storage.
+     * @throws DukeException If invalid input.
+     */
     public abstract void execute(TaskList tasks, Storage storage) throws DukeException;
 
+    /**
+     * Returns true if it is an ExitCommand.
```
Comment (seahlynn) [inline src/main/java/duke/command/Command.java]: Maybe you could mention what returning true would do to the program? To make clearer the use of this method!

**Review by seahlynn — COMMENTED** [review]
The descriptions for the different objects/constructors are clear and concise! Goooood job!!

File: src/test/java/duke/task/TodoTest.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+public class TodoTest {
+
+    @Test
+    public void testStringConversion() {
+        assertEquals("[T][✘] run", new Todo("run").toString());
+    }
+
+    @Test
+    public void markAsDone_regularTask_success() {
```
Comment (seahlynn) [inline src/test/java/duke/task/TodoTest.java]: clear naming of test function!! (sorry I just keep complimenting your code but your functions and variables are always really well-named and I am very impressed!!!)

**Review by seahlynn — COMMENTED** [review]
Maybe you could also write tests for Parser and Storage classes to make sure the parsing and loading/saving of tasks works well? Because I feel these are pretty major classes that can greatly affect the functionality of the main code!

File: src/main/java/duke/parser/Parser.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+    private static final String ERROR_INVALID_INPUT = "I'm sorry, but I don't know what that means :-(";
+    private static final String ERROR_MISSING_DESCRIPTION = "The description cannot be empty.";
+    private static final String ERROR_MISSING_TASK_ID = "The id of the task must be provided.";
+    private static final String ERROR_INVALID_TASK_ID = "The id of the task must be a number. e.g. done 1";
+    private static final String ERROR_TOO_MANY_ARGUMENTS = "There are too many arguments for this command.";
+
+    public static Command parse(String fullCommand) throws DukeException {
+        String[] line = fullCommand.split(" ", 2);
+        switch (line[0]) {
```
Comment (seahlynn) [inline src/main/java/duke/parser/Parser.java:29]: Nice use of switch-case to prevent "arrowhead style code"!! (Not too many nested if-else loops)

File: src/main/java/duke/command/DeleteCommand.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        this.taskId = taskId;
+    }
+
+    public void execute(TaskList tasks, Storage storage) throws DukeException {
+        if (taskId < 1 || taskId > tasks.size()) {
+            throw new DukeException(ERROR_INVALID_TASK_ID);
+        }
+        Task task = tasks.get(taskId - 1);
+        tasks.remove(taskId - 1);
+        Ui.printIndented(String.format(MESSAGE_DELETE, task.toString(), tasks.size(),
```
Comment (seahlynn) [inline src/main/java/duke/command/DeleteCommand.java]: The string to be printed is quite long which makes it hard to follow. It might be better to break it up! Eg. whether to print "tasks/task" could be stored in a temporary variable.

**Review by seahlynn — COMMENTED** [review]
Code is uncluttered and logically structured! Very easy to followww vv good!!!

File: src/main/java/Duke.java [hunk 9]
```diff
@@ -244,7 +249,7 @@ private void deleteTask(int id) {
     private void saveTasks() throws DukeException {
         try {
             FileWriter fw = new FileWriter(SAVE_LOCATION);
-            for (Task task : this.tasks) {
+            for (Task task : tasks) {
                 if (task instanceof Deadline) {
                     fw.append(String.format("D | %d | %s | %s\n",
```
Comment (seahlynn) [inline src/main/java/Duke.java]: Similarly, quite a long and complicated String! I think it might be better to break it up into temporary variables instead!!

File: src/main/java/Duke.java [hunk 10]
```diff
@@ -58,18 +60,14 @@ private void run() {
...
-                    }
-                    int taskId = -1;
-                    try {
-                        taskId = Integer.parseInt(line[1]);
-                    } catch (NumberFormatException ex) {
-                        throw new DukeException(ERROR_INVALID_TASK_ID);
-                    }
-                    doTask(taskId);
+                case "done": {
+                    doTask(getId(line));
```
Comment (seahlynn) [inline src/main/java/Duke.java]: the getID function helps to simplify this line of code!! Very niceeeeeeee

**Review by seahlynn — COMMENTED** [review]
I don't know if it is necessary for all the error messages to be displayed as final variables, but it does help to tidy up the code I think i also want to do something like this HAHAHA

File: src/main/java/duke/command/AddCommand.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    public AddCommand(String type, String description) {
+        this.type = type;
+        this.description = description;
+    }
+
+    public void execute(TaskList tasks, Storage storage) throws DukeException {
+        Task task;
+        switch (type) {
+        case "event": {
+            if (!description.matches(".+\\s/at\\s.+$")) {
```
Comment (seahlynn) [inline src/main/java/duke/command/AddCommand.java]: I find these lines quite confusing, because of all the symbols. Perhaps you could break it down to make it more readable for people who don't code as well!

File: src/main/java/duke/storage/Storage.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+                case "E":
+                    task = new Event(input[2], LocalDateTime.parse(input[3], DATE_TIME_FORMATTER),
+                            input[1].equals("1"));
+                    break;
+                default:
+                    throw new DukeException(ERROR_FAILED_TO_READ);
+                }
+                tasks.add(task);
+            }
+        } catch (FileNotFoundException | DukeException | ArrayIndexOutOfBoundsException
```
Comment (seahlynn) [inline src/main/java/duke/storage/Storage.java]: Perhaps the exceptions could be split up so that they throw DukeExceptions with different error messages to make it clearer to the user what and where the error occured!

File: src/main/java/duke/ui/Ui.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
+package duke.ui;
+
+import java.util.Scanner;
+
+public class Ui {
```
Comment (seahlynn) [inline src/main/java/duke/ui/Ui.java:3]: This class has very well-defined methods, which makes the rest of the code in the other classes very easy to read!! However, I think that the purpose of the printIndented() function could be abit confusing for those reading the code, so perhaps that could be better explained or named!

File: src/main/java/Duke.java (hunk 3 above)
Comment (jonchan51) [reply src/main/java/Duke.java]: Oh thanks for pointing this out! Did not realize that rule in the coding standards.

File: src/main/java/duke/command/AddCommand.java (hunk 11 above)
Comment (jonchan51) [reply src/main/java/duke/command/AddCommand.java]: This is actually regex so I don't think there's a way to make it more readable haha, I'll add a comment to describe what it does, should make it clearer that way.

Regarding the error messages being displayed as final variables, I think it's a matter of choice, I chose to put them as constants as I wanted them to be easily editable, so changing the message can just be done at the top of the file. I actually hope to refactor this into a separate file or class so that it is possible to edit all error and messages in one file, but not sure whether this is even possible.

File: src/test/java/duke/task/TodoTest.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
+package duke.task;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TodoTest {
+
+    @Test
+    public void testStringConversion() {
```
Comment (alxkohh) [inline src/test/java/duke/task/TodoTest.java]: Very simple JUnit tests indeed (as mentioned in your commit message), but I feel that these test methods are essential to the overall correctness of your program. Very descriptive name used for test methods, making it very readable, great job!

File: src/main/java/duke/parser/Parser.java [hunk 15]
```diff
@@ -12,6 +12,13 @@
     private static final String ERROR_INVALID_TASK_ID = "The id of the task must be a number. e.g. done 1";
     private static final String ERROR_TOO_MANY_ARGUMENTS = "There are too many arguments for this command.";
 
+    /**
+     * Returns a Command object depending on the command inputted by the User.
+     *
+     * @param fullCommand User input.
+     * @return Command object depending on given input.
+     * @throws DukeException If invalid input.
```
Comment (alxkohh) [inline src/main/java/duke/parser/Parser.java]: Good job writing all the javadocs, I think they are informative, yet kept very concise! For this method's javadoc, I think you can try to further explain what you mean by "invalid input". I feel that it is a little vague, just my two cents!

File: src/main/java/duke/command/FindCommand.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
+package duke.command;
+
+import duke.DukeException;
+import duke.storage.Storage;
+import duke.task.Task;
+import duke.task.TaskList;
+import duke.ui.Ui;
+
+public class FindCommand extends Command {
+    private String search;
```
Comment (alxkohh) [inline src/main/java/duke/command/FindCommand.java]: Overall good variable naming, but for this case, maybe you can consider using a more descriptive variable name for this String? For example, "toSearch", "keyword", "toFind".

File: src/main/java/Deadline.java [hunk 17]
```diff
@@ -2,10 +2,22 @@
...
+     * Returns deadline of task.
+     * @return Deadline of task.
+     */
+    public String getDeadline() {
+        return this.by;
+    }
+
     @Override
     public String toString() {
         return String.format("[D]%s (by: %s)", super.toString(), this.by);
```
Comment (alxkohh) [inline src/main/java/Deadline.java]: Very apt use of format specifiers throughout your project when it comes to string processing and printing. Well done!

File: src/main/java/duke/command/AddCommand.java [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+public class AddCommand extends Command {
+    private String type;
+    private String description;
+
+    private static final String MESSAGE_ADD      = "Got it. I've added this task:\n  %s\n"
+            + "Now you have %d %s in the list.";
```
Comment (alxkohh) [inline src/main/java/duke/command/AddCommand.java:21]: Great design thinking on your part to declare all this repetitive messages as constants (static final), and once again good use of format specifiers in those Strings!

**Comment by alxkohh** [comment]
Jonathan, overall I think your program is very well designed and well written. Good variable naming, good use of methods as re-usable blocks of logic, and good application of your Java knowledge, as seen from your apt use of format specifiers and declaration of static final constants. Keep up the good work!
